internal/backend/runtime/omni/controllers/omni/machine_teardown.go

```diff
@@ -0,0 +1,194 @@
+	if ptr.Type() == omni.TalosConfigType {
+		statuses, err := safe.ReaderListAll[*omni.MachineStatus](ctx, r, state.WithLabelQuery(resource.LabelEqual(omni.LabelCluster, ptr.ID())))
+		if err != nil {
+			return nil, err
+		}
+
+		return safe.ToSlice(statuses, func(s *omni.MachineStatus) resource.Pointer {
+			return s.Metadata()
+		}), nil
+	}
+
+	return nil, fmt.Errorf("unexpected resource type %q", ptr.Type())
+}
+
+// Reconcile implements controller.QController interface.
+func (ctrl *MachineTeardownController) Reconcile(ctx context.Context, logger *zap.Logger, r controller.QRuntime, ptr resource.Pointer) error {
+	machineStatus, err := safe.ReaderGetByID[*omni.MachineStatus](ctx, r, ptr.ID())
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	if machineStatus.Metadata().Phase() == resource.PhaseTearingDown {
+		if err := ctrl.resetMachine(ctx, r, machineStatus, logger); err != nil {
+			return err
+		}
+
+		return r.RemoveFinalizer(ctx, machineStatus.Metadata(), ctrl.Name())
+	}
+
+	if !machineStatus.Metadata().Finalizers().Has(ctrl.Name()) {
+		return r.AddFinalizer(ctx, machineStatus.Metadata(), ctrl.Name())
+	}
+
+	return nil
+}
+
+func (ctrl *MachineTeardownController) resetMachine(ctx context.Context, r controller.QRuntime,
+	machineStatus *omni.MachineStatus, logger *zap.Logger,
+) error {
+	client, err := ctrl.getClient(ctx, r, machineStatus)
+	if err != nil {
+		return err
+	}
+
+	disks, err := client.Disks(ctx)
+	if err != nil {
+		logger.Warn("machine wipe check failed", zap.Error(err))
+
+		return nil
+	}
+
+	var installed bool
+
+	for _, m := range disks.Messages {
+		for _, d := range m.Disks {
+			if d.SystemDisk {
+				installed = true
+
+				break
+			}
+		}
+	}
+
+	if !installed {
+		logger.Info("skipping machine wipe as Talos is not installed")
+
+		return nil
+	}
+
+	// try to wipe the machine without any attempts to retry it
+	err = client.Reset(ctx, false, false)
+	if err != nil {
+		logger.Warn("machine wipe failed", zap.Error(err))
+
+		return nil
+	}
+
+	logger.Info("wiped Talos on the machine")
+
+	return nil
+}
+
+func (ctrl *MachineTeardownController) getClient(
+	ctx context.Context,
+	r controller.QRuntime,
+	machineStatus *omni.MachineStatus,
+) (*client.Client, error) {
+	address := machineStatus.TypedSpec().Value.ManagementAddress
+	opts := talos.GetSocketOptions(address)
+
+	clusterName, ok := machineStatus.Metadata().Labels().Get(omni.LabelCluster)
+	if !ok {
+		return client.New(ctx,
+			append(
+				opts,
+				client.WithTLSConfig(insecureTLSConfig),
+				client.WithEndpoints(address),
+			)...)
+	}
+
+	talosConfig, err := safe.ReaderGet[*omni.TalosConfig](ctx, r, omni.NewTalosConfig(resources.DefaultNamespace, clusterName).Metadata())
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil, xerrors.NewTaggedf[qtransform.SkipReconcileTag]("cluster '%s' talosconfig not found: %w", clusterName, err)
```

Comment:
and if it hasn't been found... should the machine just be removed? if there's no talosconfig, there should be no Talos on it? or it's in maintenance mode?